### PR TITLE
Login flow

### PR DESCRIPTION
**Pull request by Timothy-Tolley**
This is the authentication and login flow routing 
Includes:
waitlist
github auth
user profiles 
user-profile migration

### CODE REVIEW

File: client/components/App.jsx [hunk 1]
```diff
@@ -1,28 +1,46 @@
...
+import {Route, withRouter} from 'react-router-dom'
 
 import '../styling/main.css'
 
 import Home from './Home'
 import Header from './Header'
 import SignIn from './SignIn'
-import Register from './Register'
 import Profile from './Profile'
+import Register from './Register'
```
Comment (don-smith) [inline client/components/App.jsx]: If you moved this line down for aesthetic reasons, you are _definitely_ strumming my heart strings. Thank you :heart:

File: client/components/App.jsx [hunk 2]
```diff
@@ -1,28 +1,46 @@
...
+import WaitList from './WaitList'
 import ErrorMessage from './ErrorMessage'
 
-const App = () => {
-  return (
-    <Router>
+class App extends React.Component {
+  constructor (props) {
+    super(props)
+    this.setToken = this.setToken.bind(this)
```
Comment (don-smith) [inline client/components/App.jsx]: We don't need to bind `setToken`. You can remove this. Which also means the constructor is no longer needed.

File: client/components/Profile.jsx [hunk 3]
```diff
@@ -25,7 +29,7 @@ class Profile extends React.Component {
   }
 
   render () {
-    const {username, currentPassword, newPassword, confirm, showMatch, match} = this.state
+    const {username, email, cohort, profilePic, bio, currentPassword, newPassword, confirm, showMatch, match} = this.state
```
Comment (don-smith) [inline client/components/Profile.jsx]: Let's break this long single line into at least 2 lines so we don't exceed 80 characters. If it is shorter than 80 characters already, you don't need to change it.

File: client/components/Profile.jsx [hunk 4]
```diff
@@ -34,24 +38,51 @@ class Profile extends React.Component {
               <fieldset>
                 <legend>Profile</legend>
 
-                <label htmlFor='username'>Username</label>
+                {/* <img src={profilePic} alt={name}> </img> */}
```
Comment (don-smith) [inline client/components/Profile.jsx]: Are we keeping this here for a good reason?

File: client/components/Profile.jsx [hunk 5]
```diff
@@ -34,24 +38,51 @@ class Profile extends React.Component {
...
+
+                <label htmlFor='email'>Email: </label>
+                <input id='email' name='email' placeholder='email'
+                  onChange={this.handleChange} value={email} />
+                <br/>
+
+                <label htmlFor='cohort'>Cohort: </label>
+                <input id='cohort' name='cohort' placeholder='cohort'
+                  onChange={this.handleChange} value={cohort} />
+                <br/>
```
Comment (don-smith) [inline client/components/Profile.jsx]: Cohort should only be shown on the grad profile because it isn't relevant for staff and employers. So we can remove it from here.

File: client/components/Profile.jsx [hunk 6]
```diff
@@ -34,24 +38,51 @@ class Profile extends React.Component {
               <fieldset>
                 <legend>Profile</legend>
```
Comment (don-smith) [inline client/components/Profile.jsx]: Let's call this "User Profile"

File: client/components/Profile.jsx [hunk 7]
```diff
@@ -34,24 +38,51 @@ class Profile extends React.Component {
...
+                <br/>
+
+                <label htmlFor='profilePic'>Upload Profile Picture: </label>
+                <input id='profilePic' name='profilePic' placeholder='profile pic url here'
+                  onChange={this.handleChange} value={profilePic} />
+                <br/>
+
+                <label htmlFor='bio'>Bio: </label>
+                <input id='bio' name='bio' placeholder='bio'
+                  onChange={this.handleChange} value={bio} />
```
Comment (don-smith) [inline client/components/Profile.jsx]: This should be a `<textarea>`

File: client/components/Profile.jsx [hunk 8]
```diff
@@ -34,24 +38,51 @@ class Profile extends React.Component {
               <fieldset>
                 <legend>Profile</legend>
 
-                <label htmlFor='username'>Username</label>
+                {/* <img src={profilePic} alt={name}> </img> */}
+
+                <label htmlFor='username'>Username: </label>
                 <input id='username' name='username' placeholder='username'
                   onChange={this.handleChange} value={username} />
+                <br/>
```
Comment (don-smith) [inline client/components/Profile.jsx]: Let's remove all of these line breaks. Let's use CSS styling for formatting instead.

File: client/components/Profile.jsx [hunk 9]
```diff
@@ -78,8 +109,8 @@ class Profile extends React.Component {
 
   handleSubmit (e) {
     const {id, updateProfile} = this.props
-    const {username, currentPassword, newPassword, confirm} = this.state
-    updateProfile(id, username, currentPassword, newPassword, confirm)
+    const {username, email, cohort, profilePic, bio, currentPassword, newPassword, confirm} = this.state
+    updateProfile(id, username, email, cohort, profilePic, bio, currentPassword, newPassword, confirm)
```
Comment (don-smith) [inline client/components/Profile.jsx]: Let's pass an object with multile properties into `updateProfile` instead of all these input parameters.

File: client/components/Profile.jsx [hunk 10]
```diff
@@ -78,8 +109,8 @@ class Profile extends React.Component {
 
   handleSubmit (e) {
     const {id, updateProfile} = this.props
-    const {username, currentPassword, newPassword, confirm} = this.state
-    updateProfile(id, username, currentPassword, newPassword, confirm)
+    const {username, email, cohort, profilePic, bio, currentPassword, newPassword, confirm} = this.state
```
Comment (don-smith) [inline client/components/Profile.jsx]: If this is more than 80 characters, let's break it up across 2 lines.

File: client/components/Profile.jsx [hunk 11]
```diff
@@ -95,10 +126,10 @@ function mapStateToProps ({userDetails}) {
 
 function mapDispatchToProps (dispatch) {
   return {
-    updateProfile: (id, username, currentPassword, newPassword, confirm) => {
+    updateProfile: (id, username, email, cohort, profilePic, bio, currentPassword, newPassword, confirm) => {
```
Comment (don-smith) [inline client/components/Profile.jsx]: See comments above about input parameters for `updateProfile` and line length.

File: client/components/SignIn.jsx [hunk 12]
```diff
@@ -22,7 +22,13 @@ class SignIn extends React.Component {
       <div className='sign-in'>
         <section className='section'>
           <div className='content'>
-            <h1>Sign in</h1>
+            <h1>Student Sign-in</h1>
+            <button className='button is-primary'><a href='http://localhost:3000/auth/github'>Continue with GitHub</a></button>
```
Comment (don-smith) [inline client/components/SignIn.jsx]: Let's put the `<a>` on its own line.
We also need to use an environment variable in `.env` for the host so we only use `http://localhost:3000` while we're in a development (not production).

File: client/components/WaitList.jsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import React from 'react'
+import {connect} from 'react-redux'
+
+// import db from '../../server/db/users'
+import {getAllUsers} from '../actions/auth'
+// import {showError, clearError} from '../actions/error'
```
Comment (don-smith) [inline client/components/WaitList.jsx]: Delete comments

File: client/components/WaitList.jsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+                  <td> {user.username} </td>
+                  <td>
+                    <input type='checkbox' />
+                  </td>
+                </tr>
+              )
+            })
+            }
+          </tbody>
+
```
Comment (don-smith) [inline client/components/WaitList.jsx]: remove blank line

File: client/components/WaitList.jsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+//   // updateProfile: (id, username, currentPassword, newPassword, confirm) => {
+//   //   if (newPassword === confirm) {
+//   //     dispatch(clearError())
+//   //     return dispatch(updateProfile({id, username, currentPassword, newPassword}))
+//   //   }
+//   // //   dispatch(showError('New password and confirmation don\'t match'))
+//   // }
+// }
+
+// export default connect(mapStateToProps, mapDispatchToProps)(WaitList)
```
Comment (don-smith) [inline client/components/WaitList.jsx]: remove comments

File: client/components/registration/Students.jsx [hunk 16]
```diff
@@ -1,148 +1,18 @@
 import React from 'react'
-import PropTypes from 'prop-types'
+// import PropTypes from 'prop-types'
```
Comment (don-smith) [inline client/components/registration/Students.jsx]: remove

File: client/components/registration/Students.jsx [hunk 17]
```diff
@@ -1,148 +1,18 @@
...
-    register: (username, password, confirm) => {
-      if (password === confirm) {
-        dispatch(clearError())
-        return dispatch(register({username, password}))
-      }
-      dispatch(showError('Password and confirmation don\'t match'))
-    }
-  }
+// import {register} from '../../actions/auth'
+// import {showError, clearError} from '../../actions/error'
```
Comment (don-smith) [inline client/components/registration/Students.jsx]: remove comments

File: client/components/registration/Students.jsx [hunk 18]
```diff
@@ -1,148 +1,18 @@
...
+
+const Students = (props) => {
+  return (
+    <form>
+      <div className='field'>
+        <button className='button is-primary GHlogin'><a href='http://localhost:3000/auth/github'>Continue with GitHub</a></button>
```
Comment (don-smith) [inline client/components/registration/Students.jsx]: * put the `<a>` on its own line
* see comment above about localhost:3000
* rename `GHlogin` CSS class to `gh-login`

File: client/reducers/all-users.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
+import {
+  RECEIVE_ALL_USERS,
+  LOG_OFF,
+  REQUEST_ALL_USERS
+} from '../actions/auth'
```
Comment (don-smith) [inline client/reducers/all-users.js]: we can probably fit these 5 lines on a single line and still not exceed 80 characters

File: client/reducers/all-users.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+const allUsers = (state = null, action) => {
```
Comment (don-smith) [inline client/reducers/all-users.js]: The default state should be `[]` instead of null. This is also true for the return statements below (in this reducer). We should also have a unit test for each of the actions this reducer manages.

File: package.json [hunk 21]
```diff
@@ -12,7 +12,8 @@
   "scripts": {
     "assert": "exit 0",
     "build": "webpack --config ./client/webpack.config.js",
-    "dev": "nodemon --watch server & yarn build --watch",
+    "webpack:watch": "webpack --watch --config ./client/webpack.config.js",
+    "dev": "nodemon --watch server & yarn webpack:watch",
```
Comment (don-smith) [inline package.json]: I've played around with these settings and the `--watch` should work the way it was. Let's undo this change and explore further if it isn't working the way we expect.

File: server/db/migrations/20180223115953_isApproved.js [hunk 22]
```diff
@@ -0,0 +1,11 @@
+exports.up = (knex, Promise) => {
+  return knex.schema.table('users', (table) => {
+    table.boolean('isApproved')
```
Comment (don-smith) [inline server/db/migrations/20180223115953_isApproved.js]: Table columns should be snake/train_case. Let's rename to `is_approved` (for the `dropColumn` call below as well). This will likey require code changes elsewhere also.

File: server/routes/users.js [hunk 23]
```diff
@@ -10,6 +10,17 @@ router.use(bodyParser.json())
 
 module.exports = router
 
+// GET /users
+router.get('/', token.decode, (req, res) => {
```
Comment (don-smith) [inline server/routes/users.js]: Let's be sure we add a unit test for this route.

File: server/db/users.js [hunk 24]
```diff
@@ -35,6 +36,12 @@ function createUser (username, password, conn, ghid = null) {
     })
 }
 
+function getAllUsers (conn) {
```
Comment (don-smith) [inline server/db/users.js]: Needs a unit test.

File: client/components/WaitList.jsx [hunk 25]
```diff
@@ -0,0 +1,67 @@
+import React from 'react'
+import {connect} from 'react-redux'
+
+// import db from '../../server/db/users'
+import {getAllUsers} from '../actions/auth'
+// import {showError, clearError} from '../actions/error'
+
+class WaitList extends React.Component {
```
Comment (don-smith) [inline client/components/WaitList.jsx]: Let's add an Enzyme test for this component.

File: client/actions/auth.js [hunk 26]
```diff
@@ -127,6 +141,20 @@ export function getUserDetails (userId) {
   }
 }
 
+export function getAllUsers () {
```
Comment (don-smith) [inline client/actions/auth.js]: This should have a unit test. Feel free to ask for my help if you need it (you might).

File: client/actions/auth.js [hunk 27]
```diff
@@ -58,6 +60,18 @@ const receiveUserDetails = (userDetails) => {
   }
 }
 
+const receiveAllUsers = (allUsers) => {
```
Comment (don-smith) [inline client/actions/auth.js]: Needs unit test.

File: client/actions/auth.js [hunk 28]
```diff
@@ -58,6 +60,18 @@ const receiveUserDetails = (userDetails) => {
...
+  return {
+    type: RECEIVE_ALL_USERS,
+    allUsers
+  }
+}
+const requestAllUsers = () => {
```
Comment (don-smith) [inline client/actions/auth.js]: Needs unit test.

File: client/components/Profile.jsx (hunk 4 above)
Comment (JudyPuff) [reply client/components/Profile.jsx]: Yes, for users to upload images (eventually)

File: client/components/Profile.jsx (hunk 5 above)
Comment (JudyPuff) [reply client/components/Profile.jsx]: This has been deleted in the latest changes (which we weren't able to push)

File: client/components/Profile.jsx (hunk 8 above)
Comment (JudyPuff) [reply client/components/Profile.jsx]: Yes boss

**Comment by Timothy-Tolley** [comment]
Ready to be merged into Dev, Testing Is WIP
